draft-ietf-quic-transport.md

```diff
@@ -870,44 +870,109 @@ different IP or port at either endpoint, due to NAT rebinding or mobility, as
 described in {{migration}}.  Finally a connection may be terminated by either
 endpoint, as described in {{termination}}.
 
+
+## Matching Packets to Connections {#packet-handling}
+
+Incoming packets are classified on receipt.  Packets can either be associated
+with an existing connection, be discarded, or - for servers - potentially create
+a new connection.
+
+Packets that can be associated with an existing connection are handled according
+to the current state of that connection.  Packets are associated with existing
+connections using connection ID if it is present; this might include connection
+IDs that were advertised using NEW_CONNECTION_ID ({{frame-new-connection-id}}).
+Packets without connection IDs and long-form packets for connections that have
+incomplete cryptographic handshakes are associated with an existing connection
+using the tuple of source and destination IP addresses and ports.
+
+A packet that uses the short header could be associated with an existing
+connection with an incomplete cryptographic handshake.  Such a packet could be a
+valid packet that has been reordered with respect to the long-form packets that
+will complete the cryptographic handshake.  This might happen after the final
+set of cryptographic handshake messages from either peer.  These packets are
+expected to be correlated with a connection using the tuple of IP addresses and
+ports.  Packets that might be reordered in this fashion SHOULD be buffered in
+anticipation of the handshake completing.
+
+0-RTT packets might be received prior to a Client Initial packet at a server.
+If the version of these packets is acceptable to the server, it MAY buffer these
+packets in anticipation of receiving a reordered Client Initial packet.
```

Comment:
This is redundant and a bit confusing. The earlier para says "Such a packet could be a valid packet that has been reordered with respect to the long-form packets that will complete the cryptographic handshake", which applies to 0-RTT packets as well. I think the para above is adequate, and this one about 0-RTT is redundant.

The para above says reordered packets SHOULD be buffered, this one says MAY.  Either is fine with me, I'd lean towards a SHOULD.

Reply:
0-RTT uses the long header.  The last paragraph talks about short headers.  Though they are related, these are completely different conditions.

The requirements have to be different.  For packets that are reordered over the last flight of the handshake, you already have an assumption that the connection is going to be made and the corresponding state commitment.  So buffering isn't a huge burden.

For 0-RTT, you receive random 0-RTT packets from some unspecified endpoints prior to any handshake packets.  Buffering of those packets is risky - they might never be followed by a handshake packet.